### ISSUE TITLE
Run performance tests against vanilla Postgres too for comparison

### ISSUE THREAD

**Issue by hlinnaka** [zenithdb/zenith#553]
First attempt here: https://github.com/zenithdb/zenith/pull/523. But need to make it less tedious to write the baseline Postgres tests, and present the results side-by-side.

**Comment by knizhnik** [comment]
pgbench -c 10 -N  -s 1000 results:
zenith@notebook: 5k TPS
zenith@ec2 storage node: 8k TPS 
vanilla@ec2 storage node: 25k TPS

**Comment by bojanserafimov** [comment]
Here are the current results, running release build on my laptop with shared_buffers=1MB.

```
| test                                     | zenith (sec) | vanillla (sec) | zenith/vanilla |
| test_bulk_insert.insert                  |       18.175 |          6.293 |           2.89 |
| test_copy.delete                         |        3.657 |          1.348 |           2.71 |
| test_small_seqscans.run                  |       18.786 |         10.724 |           1.75 |
| test_write_amplification.run             |       18.913 |         11.034 |           1.71 |
| test_pgbench.init                        |        3.274 |           2.56 |           1.28 |
| test_gist_buffering_build.build          |       25.272 |           19.9 |           1.27 |
| test_copy.copy_reuse                     |        1.911 |          2.251 |           0.85 |
| test_copy.copy_extend                    |        2.023 |          2.393 |           0.85 |
| test_parallel_copy_same_table.load       |        25.88 |         47.353 |           0.55 |
| test_copy.vacuum                         |        0.911 |          1.732 |           0.53 |
| test_parallel_copy_different_tables.load |       18.432 |         37.379 |           0.49 |
| test_pgbench.5000_xacts                  |        4.533 |         20.536 |           0.22 |
```

I think more can be done here. Ideallly our perf tests should be an accurate representation of what's important to improve, so that we can easily compare the performance of different solutions, and prioritize our efforts well. Some next steps to get us there:
- [ ] Choose a fair value of shared_buffers for each test
- [ ] Plot this zenith/vanilla column in grafana (or [here](https://zenithdb.github.io/zenith-perf-data/))
- [ ] (out of scope for this issue) Write a bigger version of each test (same test, bigger parameters). The small version is useful as regression test in CI, while the big version is an accurate representation of real performance. We can run these big tests only when there's a need, or maybe when the small test suggests we would see a difference. Either way, persisting the parameters in the repo is valuable.

**Comment by knizhnik** [comment]
I did some more investigations of performance comparison of Zenith vs. Vanilla which can clarify bottlenecks and future directions for performance improvement. The results were produced at my notebook with SSD much slower than server SSD. I use configuration without safekeepers to exclude its disk IO.

1. Write performance. 
    I want to estimate maximal possible speed at which pageserver is able to write data.
    GC and checkpointer are disabled. Autovacuum turned off. So pageserver is mostly scattering received WAL records  

1.1 pgbench -i -s 1000
     pgbench is using COPY for populating data which write FPI. 
     wal size: 19Gb
     storage size: 14Gb
     time: 114 sec
     write speed: 125Mb/sec
     CPU usage: 
           postgres: 100%
           pageserver: ~30%
           
    So looks like in this case performance is mostly limited by COPY processing or disk write speed.
    
1.2. create table t(pk bigint, val text default repeat(' ', 256));
       insert into t (pk) values (generate_series(1,100000000));

       In this case large table (~30Gb logical size) is created using inserts.
       storage size: 35Gb
       time: 411 sec               
       write speed: 83Mb/sec
       CPU usage:
            postgres: ~50%
            pageserver: ~70%
            write speed: 85Mb/sec
    
     So in this case speed is more limited by pageserver which has to do more processing. 

1.3  The same as 1.2 but checkpointer is enabled. Insertion is completed even faster: 399 seconds, but checkpointer finish its work only after 17 minutes.
    
      time: 999 sec
      storage size: 66Gb
      write speed: 68Mb/sec

2. Read performance. 
    I am reading 't' table created above.  I have inserted loop through all relation pages in pageserver itself, to eliminate   roundtrip overhead. So I am measuring maximal possible speed of extracting pages by pageserver itself.

2.1. No image layers. With checkpointer disabled there are no image layers, so pageserver has to reconstruct each page using wal_redo postgres.

    time: 220 sec
    read speed: 160Mb/sec

2.2. With image layers

    time: 29 sec
   read speed: ~1Gb/sec

3. Vanilla 
   insertion time: 889 sec (488 with fsync=off)
   read time (select count(*)):
       sequential plan: 21 seconds
       parallel plan: 15 seconds

  So zenith without safekeeper and checkpointer is inserting data even faster than vanilla.
  But in this case read speed is about 10 times slower than with materialized pages.
  With image layers read speed is comparable with  Vanilla, which means that ideal prefetch/bulk-loading mechanism should   
  eliminate ~10x performance difference with vanilla. But it is possible only for seqscan (OLAP queries).

**Comment by bojanserafimov** [comment]
> Here are the current results, running release build on my laptop with shared_buffers=1MB.
> 
> ```
> | test                                     | zenith (sec) | vanillla (sec) | zenith/vanilla |
> | test_bulk_insert.insert                  |       18.175 |          6.293 |           2.89 |
> | test_copy.delete                         |        3.657 |          1.348 |           2.71 |
> | test_small_seqscans.run                  |       18.786 |         10.724 |           1.75 |
> | test_write_amplification.run             |       18.913 |         11.034 |           1.71 |
> | test_pgbench.init                        |        3.274 |           2.56 |           1.28 |
> | test_gist_buffering_build.build          |       25.272 |           19.9 |           1.27 |
> | test_copy.copy_reuse                     |        1.911 |          2.251 |           0.85 |
> | test_copy.copy_extend                    |        2.023 |          2.393 |           0.85 |
> | test_parallel_copy_same_table.load       |        25.88 |         47.353 |           0.55 |
> | test_copy.vacuum                         |        0.911 |          1.732 |           0.53 |
> | test_parallel_copy_different_tables.load |       18.432 |         37.379 |           0.49 |
> | test_pgbench.5000_xacts                  |        4.533 |         20.536 |           0.22 |
> ```
> 
> I think more can be done here. Ideallly our perf tests should be an accurate representation of what's important to improve, so that we can easily compare the performance of different solutions, and prioritize our efforts well. Some next steps to get us there:
> 
>     * [ ]  Choose a fair configuration of vanilla for each test. Make sure the lack of safekeepers in test setup doesn't give us advantage (pageserver doesn't fsync as often)
> 
>     * [ ]  Plot this zenith/vanilla column in grafana (or [here](https://zenithdb.github.io/zenith-perf-data/))
> 
>     * [x]  (out of scope for this issue) Write a bigger version of some tests (same test, bigger parameters). The small version is useful as regression test in CI, while the big version is an accurate representation of real performance. We can run these big tests only when there's a need, or maybe when the small test suggests we would see a difference. Either way, persisting the parameters in the repo is valuable.

Spilling the remaining tasks here https://github.com/zenithdb/zenith/issues/1299
